### PR TITLE
CSHARP-4763: Support client side projections.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/AggregateOptions.cs [hunk 1]
```diff
@@ -37,6 +37,15 @@ public class AggregateOptions
...
         private bool? _useCursor;
 
+        // implicit conversions
+        /// <summary>
+        /// Creates an AggregateOptions instance with the specified translation options.
+        /// </summary>
+        /// <param name="translationOptions">The translation options.</param>
+        /// <returns>An AggregateOptions instance.</returns>
+        public static implicit operator AggregateOptions(ExpressionTranslationOptions translationOptions) =>
+            new AggregateOptions { TranslationOptions = translationOptions };
```
Comment (rstam) [inline src/MongoDB.Driver/AggregateOptions.cs:47]: Does this implicit conversion seem OK?

It makes it much easier to create `AggregateOption` when the only setting you care about is translation options.

File: src/MongoDB.Driver/ClientSideProjectionDeserializer.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            Func<TInput, TProjection> projector)
+        {
+            _inputSerializer = Ensure.IsNotNull(inputSerializer, nameof(inputSerializer));
+            _projector = Ensure.IsNotNull(projector, nameof(projector));
+        }
+
+        /// <inheritdoc/>
+        public override TProjection Deserialize(BsonDeserializationContext context, BsonDeserializationArgs args)
+        {
+            var document = _inputSerializer.Deserialize(context);
```
Comment (rstam) [inline src/MongoDB.Driver/ClientSideProjectionDeserializer.cs:68]: Currently bringing the whole document down to the client.

A future optimization (different JIRA ticket and different PR) will figure out what pieces of the document are actually needed by the client side projection.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToPipelineTranslators/SelectMethodToPipelineTranslator.cs [hunk 3]
```diff
@@ -33,17 +35,28 @@ public static AstPipeline Translate(TranslationContext context, MethodCallExpres
...
                 if (selectorLambda.Body == selectorLambda.Parameters[0])
                 {
                     return pipeline; // ignore identity projection: Select(x => x)
                 }
 
-                var selectorTranslation = ExpressionToAggregationExpressionTranslator.TranslateLambdaBody(context, selectorLambda, sourceSerializer, asRoot: true);
-                var (projectStage, projectionSerializer) = ProjectionHelper.CreateProjectStage(selectorTranslation);
-                pipeline = pipeline.AddStages(projectionSerializer, projectStage);
+                // check for client side projection after handling identity projection
+                ClientSideProjectionHelper.ThrowIfClientSideProjection(expression, pipeline, method);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToPipelineTranslators/SelectMethodToPipelineTranslator.cs:46]: We have to throw this exception *outside* of the try/catch below.

File: tests/MongoDB.Driver.TestHelpers/DriverTestConfiguration.cs [hunk 4]
```diff
@@ -147,6 +147,7 @@ public static MongoClient CreateClient(
                 ? CoreTestConfiguration.ConnectionStringWithMultipleShardRouters.ToString()
                 : CoreTestConfiguration.ConnectionString.ToString();
             var clientSettings = MongoClientSettings.FromUrl(new MongoUrl(connectionString));
+            clientSettings.TranslationOptions = new ExpressionTranslationOptions { EnableClientSideProjections = true };
```
Comment (rstam) [inline tests/MongoDB.Driver.TestHelpers/DriverTestConfiguration.cs:150]: I defaulted to true for running the test suite in order to find all the tests that needed to be refactored after adding support for client side projections.

The default for `EnableClientSideProjections` is `false` (so actually the default is `null` but we interpret `null` as `false`).

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3529Tests.cs [hunk 5]
```diff
@@ -148,18 +150,35 @@ public void Bottom_find_the_bottom_score_across_multiple_games_example_using_Gro
...
-            notSupportedException.Message.Should().Contain("Bottom can only be used as an accumulator with GroupBy");
+            if (enableClientSideProjections)
+            {
+                var stages = Translate(collection, queryable, out var serializer);
+                AssertStages(stages, Array.Empty<string>());
+                serializer.Should().BeOfType<ClientSideProjectionDeserializer<DocumentWithArray, int>>();
+
+                var exception = Record.Exception(() => queryable.ToList());
+                exception.Should().BeOfType<NotSupportedException>();
+                exception.Message.Should().Contain("This method is not functional");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3529Tests.cs:174]: This error message is not very helpful.

I would reword it to something like:

```The Bottom method is intended to be used in LINQ queries executed on the server and cannot be used in a client-side projection```

There are *many* methods with the same error message right now and they should probably all be reworded. Not sure if as part of this ticket or later as part of a separate ticket.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3958Tests.cs [hunk 6]
```diff
@@ -139,60 +140,126 @@ public void Sort_on_mixed_type_fields_example_should_work()
...
-            invalidOperationException.Message.Should().Contain("ThenBy and ThenByDescending cannot be used when OrderBy or OrderByDescending is sorting on the entire object");
+            if (enableClientSideProjections)
+            {
+                var stages = Translate(collection, queryable, out var outputSerializer);
+                AssertStages(stages, Array.Empty<string>());
+                outputSerializer.Should().BeAssignableTo<IClientSideProjectionDeserializer>();
+
+                var result = queryable.Single();
+                var exception = Record.Exception(() => result.Result.ToList());
+                exception.Should().BeOfType<InvalidOperationException>(); // TeamMembers are not IComparable
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3958Tests.cs:164]: This `Select` *could* have been executed client side *if* the `TeamMember` class had implemented `IComparable`.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3958Tests.cs [hunk 7]
```diff
@@ -139,60 +140,126 @@ public void Sort_on_mixed_type_fields_example_should_work()
...
-            var invalidOperationException = exception.Should().BeOfType<ExpressionNotSupportedException>().Subject;
-            invalidOperationException.Message.Should().Contain("ThenBy and ThenByDescending cannot be used to sort on the entire object");
+            if (enableClientSideProjections)
+            {
+                var stages = Translate(collection, queryable, out var outputSerializer);
+                AssertStages(stages, Array.Empty<string>());
+                outputSerializer.Should().BeAssignableTo<IClientSideProjectionDeserializer>();
+
+                var result = queryable.Single();
+                result.Result.Select(tm => tm.Name).Should().Equal("Charlie", "Dallas", "Pat");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3958Tests.cs:225]: Even though `TeamMember` does not implement `IComparable` we got away with it here because there are no ties when sorting on `Name`.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4100ExpressionTests.cs [hunk 8]
```diff
@@ -74,30 +75,54 @@ public void Contains_with_string_constant_and_char_field_represented_as_string_s
...
+            var queryable = collection.AsQueryable(translationOptions)
+                .Select(x => new { R = x.S.Contains(x.CC) });
 
-            exception.Should().BeOfType<ExpressionNotSupportedException>();
-            exception.Message.Should().Contain("it is not serialized as a string");
+            if (enableClientSideProjections)
+            {
+                var stages = Translate(collection, queryable, out var serializer);
+                AssertStages(stages, Array.Empty<string>());
+                serializer.Should().BeAssignableTo<IClientSideProjectionDeserializer>();
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4100ExpressionTests.cs:93]: The collection doesn't have any data (yet) so we can't execute the query and assert against the results.

Should I add data and refactor a whole bunch of tests to execute the queries with client side projections and assert against the results?

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4410Tests.cs [hunk 9]
```diff
@@ -174,94 +175,178 @@ public void Select_with_enum_comparison_should_work(int i, string projectionAsSt
...
+            if (enableClientSideProjections)
+            {
+                var stages = Translate(collection, queryable, out var outputSerializer);
+                AssertStages(stages, Array.Empty<string>());
+                outputSerializer.Should().BeAssignableTo<IClientSideProjectionDeserializer>();
 
-            exception.Should().BeOfType<ExpressionNotSupportedException>();
-            exception.Message.Should().Contain("because the two enums being compared are serialized using different serializers");
+                var results = queryable.ToList();
+                results.Should().Equal(true, true, true, true, true, true);
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4410Tests.cs:197]: This is an example of a projection that can't be executed server side (because the two enums being compared are represented differently in the database) but can be executed client side because even though the two enums are represented differently in the database by the time they have been deserialized client side they are just enum values that can be compared.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4517Tests.cs [hunk 10]
```diff
@@ -42,24 +43,41 @@ public void Filter_with_comparison_of_different_types_should_throw()
...
+            if (enableClientSideProjections)
+            {
+                var stages = Translate(collection, queryable, out var outputSerializer);
+                AssertStages(stages, Array.Empty<string>());
+                outputSerializer.Should().BeAssignableTo<IClientSideProjectionDeserializer>();
 
-            exception.Should().BeOfType<ExpressionNotSupportedException>();
-            exception.Message.Should().Contain("because operand types are not compatible with each other");
+                var results = queryable.ToList();
+                results.Select(r => r.R).Should().Equal(true);
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4517Tests.cs:65]: In this case we couldn't translate the `x.Id == 1` expression to MQL because it involves a user-defined implementation of the `==` operator.

But when we execute this projection client side the user-defined implementation of the `==` operator is accessible.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4763Tests.cs [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+            {
+                var exception = Record.Exception(() => Translate(collection, aggregate));
+                exception.Should().BeOfType<ExpressionNotSupportedException>();
+                exception.Message.Should().Contain("MyFunction");
+            }
+        }
+
+        [Theory]
+        [ParameterAttributeData]
+        public void Aggregate_Project_with_client_side_projection_Match_should_throw(
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4763Tests.cs:175]: More tests like this need to be written for all the kinds of things that could follow a client-side projection.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4957Tests.cs [hunk 12]
```diff
@@ -99,16 +115,7 @@ private class C
         {
             public int Id { get; set; }
             public int X { get; set; }
-            [BsonSerializer(typeof(YSerializer))] public int Y { get; set; }
-        }
-
-        private class YSerializer : StructSerializerBase<int>
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4957Tests.cs:105]: This test didn't actually require a custom serializer, but when it was first written we hadn't implemented `Equals` for serializers and using a custom serializer of a completely different class was a way to get this test to pass.

Now it is sufficient to just use two standard `IntSerializers` configured with different representations.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Linq3IntegrationTest.cs [hunk 13]
```diff
@@ -74,12 +74,15 @@ protected IMongoDatabase GetDatabase(string databaseName = null)
...
+        protected static List<BsonDocument> Translate<TDocument, TResult>(IMongoCollection<TDocument> collection, IAggregateFluent<TResult> aggregate) =>
+            Translate(collection, aggregate, out _);
+
+        protected static List<BsonDocument> Translate<TDocument, TResult>(IMongoCollection<TDocument> collection, IAggregateFluent<TResult> aggregate, out IBsonSerializer<TResult> outputSerializer)
         {
             var pipelineDefinition = ((AggregateFluent<TDocument, TResult>)aggregate).Pipeline;
             var documentSerializer = collection.DocumentSerializer;
-            var translationOptions = collection.Database.Client.Settings.TranslationOptions;
-            return Translate(pipelineDefinition, documentSerializer, translationOptions);
+            var translationOptions = aggregate.Options?.TranslationOptions.AddMissingOptionsFrom(collection.Database.Client.Settings.TranslationOptions);
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Linq3IntegrationTest.cs:84]: This change was necessary to merge the two sets of translation options before passing them to `Translate` on the next line.

There are other similar changes that I didn't comment on each time.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NegateExpressionToAggregationExpressionTranslatorTests.cs [hunk 14]
```diff
@@ -99,14 +101,30 @@ public void Negate_decimal128_should_work()
...
+            var queryable = Queryable.Select(collection.AsQueryable(translationOptions), i => -i.DecimalAsString);
+
+            if (enableClientSideProjections)
+            {
+                var stages = Translate(collection, queryable, out var outputSerializer);
+                AssertStages(stages, Array.Empty<string>());
+                outputSerializer.Should().BeAssignableTo<IClientSideProjectionDeserializer>();
+
+                var results = queryable.ToList();
+                results.Should().Equal(10M, -5M, 0M, decimal.MaxValue, -decimal.MaxValue);
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NegateExpressionToAggregationExpressionTranslatorTests.cs:121]: Even though the `DecimalAsString` is a `string` in the database by the time we deserialize the value and pass it to the client-side projection it is a `decimal` instance.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationWithLinq2Tests/Translators/AggregateProjectTranslatorTests.cs [hunk 15]
```diff
@@ -1756,7 +1756,8 @@ private ProjectedResult<TResult> Project<TResult>(Expression<Func<Root, TResult>
             var query = __collection.AsQueryable().Select(projector);
 
             var provider = (MongoQueryProvider<Root>)query.Provider;
-            var executableQuery = ExpressionToExecutableQueryTranslator.Translate<Root, TResult>(provider, query.Expression, translationOptions: null);
+            var translationOptions = new ExpressionTranslationOptions { EnableClientSideProjections = false };
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationWithLinq2Tests/Translators/AggregateProjectTranslatorTests.cs:1759]: This is an old test that I didn't bother refactoring to run with `EnableClientSideProjections` either `false` or `true`.

**Comment by rstam** [comment]
I'm checking out the failed tests on the patch build.

**Comment by rstam** [comment]
One test was failing because .NET itself was throwing a different exception in .NET Framework 4.7.2 than in other target frameworks. I had been testing against net60.

File: tests/MongoDB.Driver.TestHelpers/DriverTestConfiguration.cs (hunk 4 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.TestHelpers/DriverTestConfiguration.cs:150]: Should we revert this before merging into the main branch? As we probably do not want to run all tests with client side projection enabled.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3529Tests.cs (hunk 5 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3529Tests.cs:174]: Should we create ticket to track this work?

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4100ExpressionTests.cs (hunk 8 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4100ExpressionTests.cs:93]: I'm OK with leaving as is.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3529Tests.cs [hunk 16]
```diff
@@ -148,18 +150,35 @@ public void Bottom_find_the_bottom_score_across_multiple_games_example_using_Gro
                 new { Id = "G2", PlayerId = new { P = "PlayerA", S = 10 } });
         }
 
-        [Fact]
-        public void Bottom_without_GroupBy_should_have_helpful_error_message()
+        [Theory]
+        [ParameterAttributeData]
+        public void Bottom_without_GroupBy_should_have_helpful_error_message(
+            [Values(false, true)] bool enableClientSideProjections)
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3529Tests.cs:156]: There are really 3 possible states: `true`, `false` and `null`. Should we test all of the possible states to make sure we defaults to `false` on `null`s?

**Review by sanych-sun — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver/ExpressionTranslationOptions.cs [hunk 17]
```diff
@@ -34,22 +39,33 @@ public override bool Equals(object obj)
             return
                 base.Equals(obj) &&
                 obj is ExpressionTranslationOptions other &&
-                CompatibilityLevel.Equals(other.CompatibilityLevel);
+                CompatibilityLevel.Equals(other.CompatibilityLevel) &&
+                EnableClientSideProjections.Equals(other.EnableClientSideProjections);
```
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Driver/ExpressionTranslationOptions.cs:43]: And wouldn't you know it adding unit tests revealed some issues (but not `NullReferenceException`).

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3529Tests.cs (hunk 16 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3529Tests.cs:156]: As per our discussion on Slack added tests including `null` in `CSharp4763.cs` only. That is sufficient.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3529Tests.cs (hunk 5 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp3529Tests.cs:174]: I think so. I'm checking with Boris.

File: src/MongoDB.Driver/AggregateOptions.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Driver/AggregateOptions.cs:47]: Are we in agreement that this implicit conversion is OK?

Comment (JamesKovacs) [reply src/MongoDB.Driver/AggregateOptions.cs:47]: I agree that implicit conversions can be troublesome, but this use LGTM.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4763Tests.cs [hunk 18]
```diff
@@ -0,0 +1,490 @@
...
+        [InlineData(true, true, null)]
+        [InlineData(true, true, false)]
+        [InlineData(true, true, true)]
+        public void Find_with_client_side_projection_ToList_should_work(
+            bool useFindOptions,
+            bool useTranslationOptions,
+            bool? enableClientSideProjections)
+        {
+            var collection = GetCollection();
+            var translationOptions = new ExpressionTranslationOptions { EnableClientSideProjections = enableClientSideProjections };
```
Comment (damieng) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4763Tests.cs:38]: This line showing as unused in my IDE.

**Review by sanych-sun — APPROVED** [review]
LGTM

**Comment by rstam** [comment]
I rebased and fixed merge conflicts and made a few small changes requested. 

Force pushed a new commit to trigger the patch build.

**Review by JamesKovacs — APPROVED** [review]
LGTM
